packages/smart-forms-renderer/src/utils/extractObservation.ts

```diff
@@ -0,0 +1,257 @@
+import type {
+  CodeableConcept,
+  Observation,
+  Questionnaire,
+  QuestionnaireItem,
+  QuestionnaireResponse,
+  QuestionnaireResponseItem,
+  QuestionnaireResponseItemAnswer
+} from 'fhir/r4';
+import { readQuestionnaireResponse, transverseQuestionnaire } from './genericRecursive';
+import { getQrItemsIndex, mapQItemsIndex } from './mapItem';
+
+/**
+ * Extract an array of Observations from a QuestionnaireResponse and its source Questionnaire.
+ * @see https://build.fhir.org/ig/HL7/sdc/extraction.html#observation-based-extraction
+ *
+ * @author Riza Nafis
+ */
+export function extractObservationBased(
+  questionnaire: Questionnaire,
+  questionnaireResponse: QuestionnaireResponse
+): Observation[] {
+  const observations: Observation[] = [];
+
+  if (
+    !questionnaire.item ||
+    questionnaire.item.length === 0 ||
+    !questionnaireResponse.item ||
+    questionnaireResponse.item.length === 0
+  ) {
+    return observations;
+  }
+
+  if (!questionnaireResponse.id) {
+    questionnaireResponse.id = generateUniqueId(
+      (questionnaireResponse.identifier && questionnaireResponse.identifier.value) || 'QR'
+    );
+  }
+
+  const qItemExtractableMap = mapQItemsExtractable(questionnaire);
+
+  return readQuestionnaireResponse(
+    questionnaire,
+    questionnaireResponse,
+    extractObservationBasedRecursive,
+    { qr: questionnaireResponse, qItemMap: qItemExtractableMap }
+  );
+}
+
+const FHIR_OBSERVATION_EXTRACT_EXTENSION =
+  'http://hl7.org/fhir/uv/sdc/StructureDefinition/sdc-questionnaire-observationExtract';
+
+const FHIR_OBSERVATION_EXTRACT_CATEGORY_EXTENSION =
+  'http://hl7.org/fhir/uv/sdc/StructureDefinition/sdc-questionnaire-observation-extract-category';
+
+function extractObservationBasedRecursive(
+  qItem: QuestionnaireItem,
+  qrItemOrItems: QuestionnaireResponseItem | QuestionnaireResponseItem[] | null,
+  extraData?: { qr: QuestionnaireResponse; qItemMap: Record<string, boolean> }
+): Observation[] | null {
+  // Process items with child items
+  const observations: Observation[] = [];
+
+  if (!extraData?.qr || !extraData?.qItemMap) return observations;
+
+  // Map qrItemOrItems into an array of qrItems
+  let qrItems: QuestionnaireResponseItem[] = [];
+  if (qrItemOrItems) {
+    if (Array.isArray(qrItemOrItems)) {
+      qrItems = qrItemOrItems;
+    } else {
+      qrItems = [qrItemOrItems];
+    }
+  }
+
+  const categories: CodeableConcept[] = [];
+
+  for (const responseItem of qrItems) {
+    if (categories.length) break;
+
+    if (responseItem.extension) {
+      const exts = responseItem.extension.filter(
+        (e) => e.url === FHIR_OBSERVATION_EXTRACT_CATEGORY_EXTENSION
+      );
+
+      for (const ext of exts) {
+        if (ext.valueCodeableConcept) {
+          categories.push(ext.valueCodeableConcept);
+        }
+      }
+    }
+  }
```

Comment:
I think the checking of `FHIR_OBSERVATION_EXTRACT_CATEGORY_EXTENSION` should be done on a Questionnaire item instead of QuestionnaireResponse item?

Can you re-verify this?

Reply:
Oh yeah, that's true. I guess I misread the logic from LForms. I'll fix it

Reply:
I already fix the issue on extract categories logic, could you check it one more time @fongsean ?

I follow this guideline from FHIR [link](https://www.hl7.org/fhir/uv/sdc/extraction.html#observation-based-extraction:~:text=The%20observation%2Dextract,on%20an%20item.)
> The observation-extract-category extension only comes into play if a determination has been made that extraction will occur. The observation-extract-category on an item will override that declared on the base questionnaire or ancestor items. If there is a desire to add multiple categories, all must be declared on an item.

Which make the logic should be run when transversing qItem

Reply:
Just went through it again, LGTM!